sdk/azidentity/cache/linux.go

```diff
@@ -0,0 +1,242 @@
+	p, err := internal.CacheFilePath(name)
+	if err != nil {
+		return nil, err
+	}
+	// the user keyring is available to all processes owned by the user whereas the user
+	// *session* keyring is available only to processes in the current session i.e. shell
+	ringID, err := unix.KeyctlGetKeyringID(unix.KEY_SPEC_USER_KEYRING, true)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get the user keyring due to error %q", err)
+	}
+	// Attempt to link a persistent keyring to the user's plain old keyring. This enables adding
+	// "persistent" keys that survive all the user's login sessions being deleted. Like all other
+	// keys, persistent keys exist only in kernel memory and are therefore lost on shutdown. If
+	// the attempt fails--some systems don't support persistent keyrings--we just use the plain
+	// old keyring.
+	if persistentRing, err := unix.KeyctlInt(unix.KEYCTL_GET_PERSISTENT, -1, ringID, 0, 0); err == nil {
+		ringID = persistentRing
+	}
+	return &keyring{description: name, file: p, ringID: ringID}, nil
+}
+
+func (k *keyring) Delete(context.Context) error {
+	if k.keyID != 0 && k.ringID != 0 {
+		_, err := unix.KeyctlInt(unix.KEYCTL_UNLINK, k.keyID, k.ringID, 0, 0)
+		if err != nil && !isKeyInvalidOrNotFound(err) {
+			return fmt.Errorf("failed to delete cache data due to error %q", err)
+		}
+	}
+	err := os.Remove(k.file)
+	if errors.Is(err, os.ErrNotExist) {
+		return nil
+	}
+	return err
+}
+
+func (k *keyring) Read(context.Context) ([]byte, error) {
+	b, err := os.ReadFile(k.file)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	if len(b) == 0 {
+		return nil, nil
+	}
+	j, err := jwe.ParseCompactFormat(b)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't parse cache data due to error %q", err)
+	}
+	plaintext, err := k.decrypt(j)
+	return plaintext, err
+}
+
+func (k *keyring) Write(_ context.Context, data []byte) error {
+	if len(data) == 0 {
+		return nil
+	}
+	j, err := k.encrypt(data)
+	if err != nil {
+		return err
+	}
+	content, err := j.Serialize()
+	if err != nil {
+		return err
+	}
+	err = os.WriteFile(k.file, []byte(content), 0600)
+	if errors.Is(err, os.ErrNotExist) {
+		err = os.MkdirAll(filepath.Dir(k.file), 0700)
+		if err == nil {
+			err = os.WriteFile(k.file, []byte(content), 0600)
+		}
+	}
+	return err
+}
+
+func (k *keyring) createKey() ([]byte, error) {
+	// allocate an extra byte because keyring payloads must have a null terminator
+	key := make([]byte, keySize+1)
+	_, err := rand.Read(key)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't create an encryption key due to error %q", err)
+	}
+	key[keySize] = 0
+	id, err := unix.AddKey(userKey, k.description, key, k.ringID)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't store encryption key due to error %q", err)
+	}
+	k.key = key[:keySize]
+	k.keyID = id
+	return k.key, nil
+}
+
+func (k *keyring) decrypt(j jwe.JWE) ([]byte, error) {
+	for tries := 0; tries < 2; tries++ {
```

Comment:
How did you decide on 2?

Reply:
I see.  Mostly I was curious if the value was selected from some targeted testing/experimentation.  But one try plus one retry makes sense.

In the event that the retry fails, what error is surfaced to the caller, and how does one recover?

Reply:
If the retry fails, it means the data can't be decrypted because its encryption key has been replaced. There's no recovery from that so `Read` returns no data and a `nil` error as if the cache doesn't exist, which is effectively the case.

Reply:
So all of this is transparent to the caller then.  And since there's effectively no cache, a new one is created then?

Reply:
Yes. The next `Write` will encrypt with the new key and overwrite the file.

Reply:
Got it, thanks for walking me through it.